server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -172,7 +177,57 @@ boolean isJoinPending() {
         return pendingOutgoingJoins.iterator().hasNext();
     }
 
-    void sendJoinRequest(DiscoveryNode destination, Optional<Join> optionalJoin) {
+    // package-private for testing
+    static class FailedJoinAttempt {
+        private final DiscoveryNode destination;
+        private final JoinRequest joinRequest;
+        private final TransportException exception;
+        private final long timestamp;
+
+        FailedJoinAttempt(DiscoveryNode destination, JoinRequest joinRequest, TransportException exception) {
+            this.destination = destination;
+            this.joinRequest = joinRequest;
+            this.exception = exception;
+            this.timestamp = System.nanoTime();
+        }
+
+        void maybeLogNow() {
+            logger.log(getLogLevel(exception),
+                    () -> new ParameterizedMessage("failed to join {} with {}", destination, joinRequest),
+                    exception);
+        }
+
+        static Level getLogLevel(TransportException e) {
+            if (e instanceof RemoteTransportException) {
+                Throwable cause = e.getCause();
+                if (cause != null &&
+                        cause instanceof CoordinationStateRejectedException  ||
+                        cause instanceof FailedToCommitClusterStateException ||
+                        cause instanceof NotMasterException) {
+                    return Level.DEBUG;
+                }
+            }
+            return Level.INFO;
+        }
+
+        void logWarnWithTimestamp() {
+            logger.info(() -> new ParameterizedMessage("last failed join attempt was {} ms ago, failed to join {} with {}",
+                            TimeValue.timeValueMillis(TimeValue.nsecToMSec(System.nanoTime() - timestamp)),
+                            destination,
+                            joinRequest),
+                    exception);
+        }
+    }
+
+
+    void logLastFailedJoinAttempt() {
+        FailedJoinAttempt attempt = lastFailedJoinAttempt;
+        if (attempt != null) {
+            attempt.logWarnWithTimestamp();
```

Comment:
should we avoid repeatedly logging the same failed join attempt? I wonder if we should set `lastFailedJoinAttempt` back to null here (and use AtomicReference)

Reply:
I think that we need to log lastFailedJoinAttempt each time we log a warn in `ClusterFormationHelper`, because that way we know that cluster still can not form and failed join attempt still could be the reason for it. Otherwise, next time `ClusterFormationHelper` logs a warning and no failed join attempt, we don't know if the issue with join is actually resolved.

Reply:
I'm with @ywelsch on this one. Imagine one join attempt fails with an exception and then the network is disconnected. We will continue to log this join exception, with a new timestamp each time, and I think that's confusing because this exception has nothing to do with the ongoing failure to form a cluster. It's true that we log how many milliseconds ago the exception occurred but I think that could easily be overlooked.

Reply:
Ok, I fixed it here https://github.com/elastic/elasticsearch/pull/39756/commits/283fdeaba7fc9a6fb20345d60ee1c1bb9aae214d